src/table/boot.rs

```diff
@@ -17,6 +17,25 @@ use core::ptr::NonNull;
 use core::{ptr, slice};
 
 /// Contains pointers to all of the boot services.
+///
+/// # Use of [`UnsafeCell`] for protocol references
+///
+/// Some protocols require mutable access to themselves. For example,
+/// most of the methods of the [`Output`] protocol take `&mut self`,
+/// because the internal function pointers specified by UEFI for that
+/// protocol take a mutable `*This` pointer. We don't want to directly
+/// return a mutable reference to a protocol though because the lifetime
+/// of the protocol is tied to `BootServices`. (That lifetime improves
+/// safety by ensuring protocols aren't accessed after exiting boot
+/// services.) If methods like [`open_protocol`] protocol took a mutable
+/// reference to `BootServices` and returned a mutable reference to a
+/// protocol it would prevent all other access to `BootServices` until
+/// the protocol reference was dropped. To work around this, the
+/// protocol reference is wrapped in an [`UnsafeCell`]. Callers can then
+/// get a mutable reference to the protocol if needed.
+///
+/// [`Output`]: crate::proto::console::text::Output
+/// [`open_protocol`]: BootServices::open_protocol
```

Comment:
Is the `BootServices` documentation the most relevant place for this information? Would it make more sense to move it to the `ScopedProtocol` struct, or to the `proto` module, or add links there to this doc section?

Reply:
Good question, my thinking with putting the doc on `BootServices` is that that is where all protocols get opened. I've updated this PR with a bit more detail on opening protocols in the BootServices docs, and added some more intra-doc links as you suggested.

Reply:
Looks great now, thanks!